src/tools/extract-conditionals.cpp

```diff
@@ -0,0 +1,991 @@
+                                        .allow_unregistered()
+                                        .run();
+
+        // split parsed options
+        auto command = std::find_if(parsed.options.begin(), parsed.options.end(), [](auto &o) {
+            return o.string_key == "command";
+        });
+        for (auto it = command; it != parsed.options.end(); ++it)
+        {
+            std::copy(it->original_tokens.begin(),
+                      it->original_tokens.end(),
+                      back_inserter(command_options));
+        }
+
+        parsed.options.erase(command, parsed.options.end());
+        po::store(parsed, vm);
+    }
+    catch (const boost::program_options::error &e)
+    {
+        throw osrm::util::exception(e.what());
+    }
+
+    if (vm.count("version"))
+    {
+        std::cout << OSRM_VERSION << std::endl;
+        std::exit(EXIT_SUCCESS);
+    }
+
+    if (vm.count("help") || command_options.empty())
+    {
+        const auto *executable = argv[0];
+        std::cout << boost::filesystem::path(executable).filename().string()
+                  << " [<global options>] <command> [<args>]\n\n"
+                     "Supported commands are:\n\n"
+                     "   cond-dump      Save conditional restrictions in CSV format\n"
+                     "   cond-check     Check conditional restrictions and save in CSV format\n"
+                     "   speed-dump     Save conditional speed limits in CSV format\n"
+                     "   speed-check    Check conditional speed limits and save in CSV format\n"
+                  << "\n"
+                  << global_options;
+        std::exit(EXIT_SUCCESS);
+    }
+
+    return command_options;
+}
+
+void ParseDumpCommandArguments(const char *executable,
+                               const char *command_name,
+                               const std::vector<std::string> &arguments,
+                               std::string &osm_filename,
+                               std::string &csv_filename)
+{
+    namespace po = boost::program_options;
+
+    po::options_description options(command_name);
+    options.add_options()("help,h", "Show this help message")(
+        "input,i",
+        boost::program_options::value<std::string>(&osm_filename),
+        "OSM input file in .osm, .osm.bz2 or .osm.pbf format")(
+        "output,o",
+        boost::program_options::value<std::string>(&csv_filename),
+        "Output conditional restrictions file in CSV format");
+
+    po::positional_options_description pos;
+    pos.add("input", 1).add("output", 1);
+
+    po::variables_map vm;
+    po::store(po::command_line_parser(arguments).options(options).positional(pos).run(), vm);
+
+    if (vm.count("help"))
+    {
+        std::cout << boost::filesystem::path(executable).filename().string() << " " << command_name
+                  << " [<args>]\n\n"
+                  << options;
+        std::exit(EXIT_SUCCESS);
+    }
+    boost::program_options::notify(vm);
```

Comment:
* `po::store`  stores parsed command line arguments into a variable map `vm`
* `vm.count("help")` returns number of keys in the variable map, so if --help or -h was in arguments `count` will return 1
* `po::notify` propagates data from the variable map `vm` to handlers in `po::options_description`, eg  if `--input arg`  was specified then "arg" will be handled by `boost::program_options::value<std::string>(&osm_filename)` that is basically a value copy, but handler can be more complicated than a copy.

Nice thing that options can come from arguments line and from configuration file like
```
variables_map vm;
store(parse_command_line(argc, argv, desc), vm);
store(parse_config_file("example.cfg", desc), vm);
notify(vm);
```